### PR TITLE
[PINS] Add Bazel support for external projects

### PR DESCRIPTION
**Pull request by bocon13**
Bazel build support allows swss-common to used
in Bazel projects; this doesn’t change/impact the
make-based build system

Submission containing materials of a third party:
Copyright Google LLC; Licensed under Apache 2.0

Co-authored-by: Akarsh Gupta akarshgupta@google.com
Co-authored-by: Ashish Singh ashishksingh@google.com
Co-authored-by: Jay Hu jiarenhu@google.com
Co-authored-by: Manali Kumar manalikumar@google.com
Co-authored-by: Richard Yu richardyu@google.com
Co-authored-by: Robert J. Halstead rhalstea@google.com
Co-authored-by: Runming Wu runmingwu@google.com
Co-authored-by: Srikishen Pondicherry Shanmugam kishanps@google.com
Co-authored-by: Vivek Ramamoorthy vivekmoorthy@google.com
Co-authored-by: Yilan Ji yilanji@google.com

TODO

- [x] Integrate this into Azure Pipelines to prevent regressions (~~ETA: this week in a separate PR~~)
- [x] Rebase after merging dependencies: #545, #547, #556

### CODE REVIEW

**Comment by bocon13** [comment]
@qiluo-msft this is related to #547

File: WORKSPACE [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    name = "com_google_absl",
+    sha256 = "b3744a4f7a249d5eaf2309daad597631ce77ea62e0fc6abffbab4b4c3dc0fc08",
+    strip_prefix = "abseil-cpp-20200923",
+    url = "https://github.com/abseil/abseil-cpp/archive/20200923.tar.gz",
+)
+
+http_archive(
+    name = "com_google_googletest",
+    sha256 = "9dc9157a9a1551ec7a7e43daea9a694a0bb5fb8bec81235d8a1e6ef64c716dcb",
+    strip_prefix = "googletest-release-1.10.0",
```
Comment (qiluo-msft) [inline WORKSPACE:15]: >googletest [](http://example.com/codeflow?start=20&length=10)

The gtest is already installed in sonic-slave container ([https://github.com/Azure/sonic-buildimage/blob/5826a5c4e751124e2552c78ef826248b49071379/sonic-slave-stretch/Dockerfile.j2#L397](https://github.com/Azure/sonic-buildimage/blob/5826a5c4e751124e2552c78ef826248b49071379/sonic-slave-stretch/Dockerfile.j2#L397)). Is this conflicting?

Comment (bocon13) [reply WORKSPACE:15]: Bazel doesn't use the system libraries and doesn't install into system, so there should be no conflict.

File: .gitignore [hunk 2]
```diff
@@ -61,3 +61,6 @@ stamp-h1
 ###############
 common/swssloglevel
 tests/tests
+
+# Bazel Build System #
+/bazel-*
```
Comment (qiluo-msft) [inline .gitignore:66]: >/ [](http://example.com/codeflow?start=0&length=1)

Remove `/` at the beginning of the line?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -41,11 +41,21 @@ stages:
         sudo apt-get install -y python
         sudo apt-get install cmake libgtest-dev
         cd /usr/src/gtest && sudo cmake . && sudo make
+        BAZELISK_VERSION=1.10.1
```
Comment (qiluo-msft) [inline azure-pipelines.yml:44]: >1.10.1 [](http://example.com/codeflow?start=25&length=6)

Better to always use the latest stable version?

File: azure-pipelines.yml [hunk 4]
```diff
@@ -41,11 +41,21 @@ stages:
...
+        ARCH=$(dpkg --print-architecture)
+        set -x
+        sudo curl -fsSL -o /usr/local/bin/bazel \
+            https://github.com/bazelbuild/bazelisk/releases/download/v${BAZELISK_VERSION}/bazelisk-linux-${ARCH}
```
Comment (bocon13) [inline azure-pipelines.yml:48]:
```suggestion
            https://github.com/bazelbuild/bazelisk/releases/latest/download/bazelisk-linux-${ARCH}
```

File: azure-pipelines.yml (hunk 3 above)
Comment (bocon13) [reply azure-pipelines.yml:44]:
```suggestion
```

Sure, we can use latest

File: .gitignore (hunk 2 above)
Comment (bocon13) [reply .gitignore:66]: The `bazel-*` directories will only be place in the repository root. The leading slash prevents git from ignoring a directory named `bazel-<somthing>` at some other point in the tree.

> The pattern hello.* matches any file or folder whose name begins with hello.. If one wants to restrict this only to the directory and not in its subdirectories, one can prepend the pattern with a slash, i.e. /hello.*; the pattern now matches hello.txt, hello.c but not a/hello.java.

**Review by qiluo-msft — DISMISSED** [review]

**Comment by rhalstea** [comment]
@qiluo-msft @bocon13 

Thanks Qi for the approval, Before merging I think we can remove the tests/fakes/BUILD, and tests/mocks/BUILD files.

**Comment by bocon13** [comment]
> Thanks Qi for the approval, Before merging I think we can remove the tests/fakes/BUILD, and tests/mocks/BUILD files.

Those files are removed. Just need to rebase after #545 is merged.

**Review by qiluo-msft — DISMISSED** [review]

**Comment by rhalstea** [comment]
@qiluo-msft @bocon13 

Thanks for the cleanup Brian. Looks like the tests are passing as well.

Qi do you mind approving and merging the changes?

**Review by qiluo-msft — APPROVED** [review]

**Comment by rhalstea** [comment]
Thanks Qi!

**Pull request merged**